libafl/src/generators/mod.rs

```diff
@@ -74,64 +71,47 @@ where
 
 #[derive(Clone, Debug)]
 /// Generates random bytes
-pub struct RandBytesGenerator<S> {
+pub struct RandBytesGenerator {
     max_size: NonZeroUsize,
-    phantom: PhantomData<S>,
 }
 
-impl<S> Generator<BytesInput, S> for RandBytesGenerator<S>
+impl<S> Generator<BytesInput, S> for RandBytesGenerator
 where
     S: HasRand,
 {
     fn generate(&mut self, state: &mut S) -> Result<BytesInput, Error> {
-        let mut size = state.rand_mut().below(self.max_size);
-        if size == 0 {
-            size = 1;
-        }
+        let size = 1 + state.rand_mut().below(self.max_size);
         let random_bytes: Vec<u8> = (0..size)
-            .map(|_| state.rand_mut().below(NonZero::new(256).unwrap()) as u8)
+            .map(|_| state.rand_mut().below(nonzero!(256)) as u8)
             .collect();
         Ok(BytesInput::new(random_bytes))
     }
 }
 
-impl<S> RandBytesGenerator<S> {
-    /// Returns a new [`RandBytesGenerator`], generating up to `max_size` random bytes.
-    ///
-    /// If you want to save one 0 check, use [`Self::from_nonzero`].
-    pub fn new(max_size: usize) -> Result<Self, Error> {
-        let Some(max_size) = NonZero::new(max_size) else {
-            return Err(Error::illegal_argument("The max_size may not be 0."));
-        };
-        Ok(Self::from_nonzero(max_size))
-    }
-
+impl RandBytesGenerator {
     /// Returns a new [`RandBytesGenerator`], generating up to `max_size` random bytes.
     #[must_use]
-    pub fn from_nonzero(max_size: NonZeroUsize) -> Self {
-        Self {
-            max_size,
-            phantom: PhantomData,
-        }
+    pub fn new(max_size: usize) -> Self {
+        // # Safety
+        // Saturating added 1 so it's always above 0
+        let max_size = unsafe { NonZeroUsize::new(max_size.saturating_add(1)).unwrap_unchecked() };
+        Self { max_size }
     }
 }
 
 #[derive(Clone, Debug)]
 /// Generates random printable characters
-pub struct RandPrintablesGenerator<S> {
+pub struct RandPrintablesGenerator {
     max_size: NonZeroUsize,
-    phantom: PhantomData<S>,
 }
 
-impl<S> Generator<BytesInput, S> for RandPrintablesGenerator<S>
+impl<S> Generator<BytesInput, S> for RandPrintablesGenerator
 where
     S: HasRand,
 {
     fn generate(&mut self, state: &mut S) -> Result<BytesInput, Error> {
-        let mut size = state.rand_mut().below(self.max_size);
-        if size == 0 {
-            size = 1;
-        }
+        let size = 1 + state.rand_mut().below(self.max_size);
```

Comment:
This is not correct. You add 1 in `new` and here you add 1 again.
So a max size of 10 will lead to `self.max_size = 11` -> below leads to a value up to 10. Here, you add 1 to `size` so worst case the result has 11 entries.

Reply:
ok